distributed/protocol/keras.py

```diff
@@ -1,8 +1,11 @@
 from __future__ import print_function, division, absolute_import
 
-from .serialize import register_serialization, serialize, deserialize
+from .serialize import dask_serialize, dask_deserialize, serialize, deserialize
 
+import keras
 
+
+@dask_serialize.register(keras.Model)
```

Comment:
This line assumes that every Keras model is an instance of `keras.Model`, and there are cases where that's not true. Taking from their "[Getting started guide on Sequential models][1]:

[1]:https://keras.io/getting-started/sequential-model-guide/

``` python
>>> from keras.datasets import mnist
>>> from keras.models import Sequential
>>> from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPooling2D
>>> from keras import backend as K
>>> import keras
>>>
>>> model = Sequential()
>>> model.add(Conv2D(32, kernel_size=(3, 3),
>>>                  activation='relu',
>>>                  input_shape=(28, 28, 1)))
>>> model.add(Conv2D(64, (3, 3), activation='relu'))
>>> model.add(MaxPooling2D(pool_size=(2, 2)))
>>> model.add(Dropout(0.25))
>>> model.add(Flatten())
>>> model.add(Dense(128, activation='relu'))
>>> model.add(Dropout(0.5))
>>> model.add(Dense(10, activation='softmax'))
>>> type(model).mro()
... [keras.engine.sequential.Sequential,
... keras.engine.training.Model,
... keras.engine.network.Network,
... keras.engine.base_layer.Layer,
... object]
>>>
>>> model.compile(loss='categorical_crossentropy',
              optimizer=opt)
>>> type(model).mro()
... [keras.engine.sequential.Sequential,
... keras.engine.training.Model,
... keras.engine.network.Network,
... keras.engine.base_layer.Layer,
... object]
```

I think the functionality implemented at the end of this file should be preserved (`for module in ... for name in ...`). I tried adding these lines back in but couldn't get it working. Here's a gist of what I have: https://gist.github.com/stsievert/7389acd62eb6f6d151286214cb51c392

Reply:
It seems to work ok for me from your example

```python
In [5]: from distributed.protocol import serialize, deserialize

In [6]: header, frames = serialize(model)
2018-08-17 06:36:00.382137: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-08-17 06:36:00.396155: I tensorflow/core/common_runtime/process_util.cc:69] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.

In [7]: model2 = deserialize(header, frames)

In [8]: model2
Out[8]: <keras.engine.sequential.Sequential at 0x7f4d62c50828>
```

I think that we're depending here on `keras.model_from_config` to work properly

Reply:
Hm. I'm still seeing `TypeError: can't pickle _thread.RLock objects` on e5792a341045366cd2149cdee85985e9deceff5d in `deserialize` (`serialize` works fine). An example of what I'm seeing is in  https://gist.github.com/stsievert/4f593614450a143bf1b342e1fdeadce9.

Reply:
Oddly that notebook rolls through fine for me.  The only difference in versions I see is that keras is 2.2.2

Reply:
Whoops, looks like I had modifications (`.dirty` in `distributed.__version__`).

I have reverted back to e5792a341045366cd2149cdee85985e9deceff5d and tested with different Keras versions. It works for Keras 2.2.2, 2.2.0, 2.1.0, but fails with 2.0.0.

Reply:
And now https://github.com/dask/distributed/issues/2183 is what I'm seeing.

Reply:
Reproduced.  Thank you.